xarray/core/dataset.py

```diff
@@ -6864,6 +6877,174 @@ def bfill(self, dim: Hashable, limit: int | None = None) -> Self:
         new = _apply_over_vars_with_dim(bfill, self, dim=dim, limit=limit)
         return new
 
+    def fill_gaps(
+        self,
+        dim: Hashable,
+        use_coordinate: bool | Hashable = True,
+        limit: (
+            None
+            | int
+            | float
+            | str
+            | pd.Timedelta
+            | np.timedelta64
+            | datetime.timedelta
+        ) = None,
+        limit_direction: LimitDirectionOptions = "both",
+        limit_area: LimitAreaOptions | None = None,
+        max_gap: (
+            None
+            | int
+            | float
+            | str
+            | pd.Timedelta
+            | np.timedelta64
+            | datetime.timedelta
+        ) = None,
+    ) -> GapMask:
+        """Fill in gaps in the data using one of several filling methods.
+        Allows for fine control on how far to extend the valid data into the gaps and the maximum size of the gaps to fill.
+
+        Parameters
+        ----------
+        dim : Hashable
+            Specifies the dimension along which to calculate gap sizes.
+        use_coordinate : bool or Hashable, default: True
+            Specifies which index to use when calculating gap sizes.
+
+            - False: a consecutive integer index is created along ``dim`` (0, 1, 2, ...).
+            - True: the IndexVariable `dim` is used.
+            - String: specifies the name of a coordinate variable to use as the index.
+
+        limit : int, float, str, pandas.Timedelta, numpy.timedelta64, datetime.timedelta, default: None
+            Maximum number or distance of consecutive NaNs to fill.
+            Use None for no limit. When interpolating along a datetime64 dimension
+            and ``use_coordinate=True``, ``limit`` can be one of the following:
+
+            - a string that is valid input for pandas.to_timedelta
+            - a :py:class:`numpy.timedelta64` object
+            - a :py:class:`pandas.Timedelta` object
+            - a :py:class:`datetime.timedelta` object
+
+            Otherwise, ``limit`` must be an int or a float.
+            If ``use_coordinates=True``, for ``limit_direction=forward`` distance is defined
+            as the difference between the coordinate at a NaN value and the coordinate of the next valid value
+            to the left (right for ``limit_direction=backward``).
+            For example, consider::
+
+                <xarray.DataArray (x: 9)>
+                array([nan, nan, nan,  1., nan, nan,  4., nan, nan])
+                Coordinates:
+                  * x        (x) int64 0 1 2 3 4 5 6 7 8
+
+            For ``limit_direction=forward``, distances are ``[nan, nan, nan, 0, 1, 2, 0, 1, 2]``.
+            To only fill gaps less than a given length,
+            see ``max_gap``.
+        limit_direction: {"forward", "backward", "both"}, default: "forward"
+            Consecutive NaNs will be filled in this direction.
```

Comment:
I think it is actually useful to have this parameter and allow to set it independently of the filing function. Then, for example, you can do things like:
```python
n = np.nan
test=xr.DataArray([1,1, n, n, n, 2, 2], dims=["x"])
test.fill_gaps('x', limit=1, limit_direction='backward').ffill('x')
```
which effectively pulls the gap left edge towards the gap right edge:
```
<xarray.DataArray (x: 7)> Size: 56B
array([ 1.,  1., nan, nan,  1.,  2.,  2.])
Dimensions without coordinates: x
```
I think there might be use cases like for this, constructing masks or advanced indexing arrays maybe.

Reply:
Hmmm, in my (probably naive) mental model of filling, this wouldn't work like this. How would we explain in words why we get the additional `1`?

Reply:
I see your point... My view on this:
- `fill_gaps()` decides **which** values to fill (here 1 step backward, i.e. the right edge of every gap)
- The following function decides **how** they are filled (`ffill` here)

Both are two independent steps.
Maybe it is clearer to use 'left'/'right' instead of 'forward'/'backward' terminology?
```
limit_side: {"left", "right", "both"}, default: "both"
    Whether to limit the filling to the left, right, or both sides of the gap
     - "left": Fill only the left (lower index) side of each gap.
     - "right": Fill only the right (higher index) side of each gap.
     - "both": Fill from both sides of each gap.
```

Reply:
I could also join this Wednesday developer meeting to get feedback on naming/API things.

Reply:
I see what you mean. I personally think that's quite confusing, but 50% chance it's just me.

Do others have a view?

I'm thinking that filling is simpler than this: it starts at the edge of a gap — and then fills the gap from the edge, which is by definition backwards or forwards (because it's already at an edge so can only go one direction)...

Reply:
(yes re Wednesday calls! Unfortunately it clashes with something I have so I find it tough to go, but others will be there!)